### PR TITLE
Fix: Dashboard password reset after leaving cluster

### LINKED ISSUE #3155: Dashboard admin password resets to default after leaving cluster
#### Environment

- OS: Ubuntu 18.04.3 LTS (AWS EC2)
- Erlang/OTP: R21/10.3.5.4
- EMQ: 3.2.7

#### Description

If the password for the admin dashboard user is changed from the default "public" via the dashboard or cli (emqx_ctl admins passwd admin newpasswd) and the server is then joined to a cluster and then leaves that cluster, the password is reset to the default (public). This is also the case of the default password is changed in /etc/emqx/plugins/emqx_dashboard.conf.

### PR DESCRIPTION
**Pull request by gsychev**
Closes #3155

### CODE REVIEW

File: lib-ce/emqx_dashboard/src/emqx_dashboard_admin.erl [hunk 1]
```diff
@@ -128,6 +128,10 @@ change_password(Username, OldPasswd, NewPasswd) when is_binary(Username) ->
     end.
 
 change_password(Username, Password) when is_binary(Username), is_binary(Password) ->
+    case binenv(default_user_username) of
+        Username -> application:set_env(emqx_dashboard, default_user_passwd, Password);
```
Comment (zmstone) [inline lib-ce/emqx_dashboard/src/emqx_dashboard_admin.erl:132]: we'll need to call (rpc) other nodes all to set the same password.

change_password sets password in db which is replicated to other nodes
if we change app-env only locally, it's an inconsistent behaviour

**Comment by zmstone** [comment]
it's probably an overkill to subscribe to all mnesia events for one change.
a plain rpc call should do?

**Comment by gsychev** [comment]
@zmstone , plain rpc call can cause inconsistencies in different cases:  
For example, when password change is initiated simultaneously from different nodes  
Or when some node loses connection during rpc call.

Solution with subscription relies on `mnesia` to resolve this conflicts. And this table is updated very rarely, so this subscription is lightweight

**Comment by zmstone** [comment]
we need to test this too:
* a node joins the cluster after the default password has been changed from other nodes.
* the node leaves the cluster
* the node has changed default password set in app-env

i.e.
app-env should sync up with the current truth (mnesia) at boot up.

**Comment by zmstone** [comment]
converted to draft since it's still WIP

File: src/emqx.erl [hunk 2]
```diff
@@ -234,7 +234,19 @@ shutdown(Reason) ->
...
+            application:stop(emqx_dashboard), %% dashboard must be started after mnesia
+            lists:foreach(fun application:start/1 , default_started_applications()),
+            application:start(emqx_dashboard);
+
+        false ->
+            lists:foreach(fun application:start/1 , default_started_applications())
+    end.
+
+is_application_running(App) ->
+    StartedApps = proplists:get_value(started, application:info()),
```
Comment (zhongwencool) [inline src/emqx.erl:248]: @zmstone Do we have a package that does not enable `emqx_dashboard`?
I used to  think the `emqx_dashboard` is always enabled.

`application:info()` return `started, running, starting,start_p_false, loaded, loading` status.
I'm a bit curious if emqx_dashboard, which was started first without mnesia started, will cause it to stay in the `starting` state.

I would suggest removing `is_application_running(emqx_dashboard)` check if we always enable dashboard.

Comment (zmstone) [reply src/emqx.erl:248]: Technically, it’s possible to disable dashboard.
It’s just as other plugins

**Comment by zmstone** [comment]
need rebase.

**Comment by zmstone** [comment]
moved to: https://github.com/emqx/emqx/pull/7518

**Pull request closed without merging**